File: src/main/scala/com/weave/core/GraphRunner.scala
package com.weave.core

import scala.annotation.tailrec
import scala.collection.immutable.Queue
import scala.util.{Failure, Success, Try}
import scala.util.chaining.*

private[core] class GraphRunner[S, U](graph: Graph[S, U]) {

  private def attemptExecuteNode(
      node: Node[S, U],
      state: S,
      onEvent: GraphEvent => Unit
  ): Try[S] = {
    onEvent(GraphEvent.NodeStarted(node.name))
    Try({
      val update = node.f(state)
      graph.reducer.reduce(state, update)
    }).tap({
      case Failure(ex) =>
        onEvent(GraphEvent.NodeFailed(node.name, ex))
      case Success(nextState) =>
        onEvent(GraphEvent.NodeCompleted(node.name))
        onEvent(GraphEvent.CheckpointCreated(node.name, nextState))
    })
  }

  @tailrec
  private def retryExecuteNode(
      node: Node[S, U],
      state: S,
      onEvent: GraphEvent => Unit,
      remainingAttempts: Int,
      previousResult: Try[S]
  ): Try[S] = {
    previousResult match {
      case success @ Success(_)                           => success
      case failure @ Failure(_) if remainingAttempts <= 1 => failure
      case Failure(_)                                     =>
        val result = attemptExecuteNode(node, state, onEvent)
        retryExecuteNode(node, state, onEvent, remainingAttempts - 1, result)
    }
  }

  private def executeNode(
      node: Node[S, U],
      state: S,
      onEvent: GraphEvent => Unit
  ): Either[GraphError.ExecutionError, S] = {
    attemptExecuteNode(node, state, onEvent)
      .pipe(previousResult => {
        node.retryPolicy match {
          case RetryPolicy.Never                      => previousResult
          case RetryPolicy.FixedAttempts(maxAttempts) =>
            retryExecuteNode(node, state, onEvent, maxAttempts, previousResult)
        }
      })
      .fold(
        ex => {
          onEvent(GraphEvent.WorkflowFailed("workflow", ex))
          Left(GraphError.RuntimeError(node.name, ex))
        },
        Right(_)
      )
  }

  def run(
      initialState: S,
      onEvent: GraphEvent => Unit = _ => ()
  ): GraphError.ExecutionError | S = {

    // traversal BFS
    def execute(
        runtimeState: RuntimeState[S],
        stateAcc: GraphState[S],
        onEvent: GraphEvent => Unit
    ): GraphError.ExecutionError | S = {
      def fail(error: GraphError.ExecutionError): GraphError.ExecutionError = {
        onEvent(GraphEvent.WorkflowFailed("workflow", error.cause))
        error
      }

      // 4. Pick a node from work-queue, execute,
      runtimeState.dequeue match {
        case Some(workItem, newRuntimeState) =>
          val result =
            executeNode(graph.nodes(workItem.nodeName), workItem.state, onEvent)
          result match {
            case Left(error)     => error
            case Right(newState) =>
              // get child(s). Only those whose edges pass condition.
              // 5. If child node is dependent on multiple node, move to pending-joins queue.
              // 6. If child node not dependent then add to work-queue.
              val nextNodes =
                if (graph.end.contains(workItem.nodeName)) List.empty
                else graph.nextNodes(workItem.nodeName, newState)

              val (nodesForPendingJoins, nodesForWorkQueue) = nextNodes
                .partition(graph.isMultipleParentNode)

              val updatedRuntimeState = newRuntimeState
                .enqueue(nodesForWorkQueue.map(n => WorkItem(n, newState)))
                .addMultipleJoinArrival(
                  nodesForPendingJoins
                    .map(n => PendingJoinInput(n, workItem.nodeName, newState))
                )

              // 7. Start the loop again. Continue till both pending-joins and work-queue are empty.
              execute(
                updatedRuntimeState,
                stateAcc.update(workItem.nodeName -> newState),
                onEvent
              )
          }

        // Check if Pending Joins.
        case None if runtimeState.isJoinPending =>
          val finishedJoins =
            RuntimeState.finishedJoins(
              runtimeState,
              nodeName => graph.parentNodes(nodeName).toSet
            )

          if (finishedJoins.isEmpty) {
            fail(GraphError.JoinDeadlock(runtimeState.pendingJoins.keySet))
          } else {
            // If all parent processed move to work-queue
            val newRuntimeState = finishedJoins.foldLeft(runtimeState) {
              case (runtimeStateAcc, nodeName) =>
                val arrivals = runtimeState.pendingJoins(nodeName)
                val joinInputs = graph.parentNodes(nodeName).map(arrivals)
                val combinedState = graph.reducer.merge(
                  joinInputs.head,
                  joinInputs.tail*
                )
                runtimeStateAcc
                  .copy(
                    workQueue = runtimeStateAcc.workQueue.enqueue(
                      WorkItem(nodeName, combinedState)
                    ),
                    pendingJoins =
                      runtimeStateAcc.pendingJoins.removed(nodeName)
                  )
            }
            execute(newRuntimeState, stateAcc, onEvent)
          }
        case None =>
          graph.end.flatMap(stateAcc.get) match {
            case Some(resultState) => resultState
            case None              =>
              fail(
                GraphError.RuntimeError(
                  "empty work queue",
                  new RuntimeException(
                    "Workflow completed without executing the configured end node."
                  )
                )
              )
          }
      }
    }
    execute(
      RuntimeState(
        Queue.from(graph.start).map(WorkItem(_, initialState)),
        Map.empty
      ),
      GraphState.empty,
      onEvent
    )
      .tap {
        case _: GraphError.ExecutionError => ()
        case _ => onEvent(GraphEvent.WorkflowCompleted("workflow"))
      }
  }
}

File: src/main/kotlin/com/weave/core/GraphRunner.kt
package com.weave.core

/** Executes a validated graph synchronously. */
public class GraphRunner<S, U> internal constructor(private val graph: Graph<S, U>) {
    /**
     * Executes the workflow and returns an explicit success or workflow failure.
     *
     * Exceptions raised by node actions, state reduction, routing predicates, and
     * branch-state merging become [RunResult.Failure]. Exceptions raised by
     * [onEvent] escape unchanged because the callback is an observer, not part of
     * workflow computation.
     */
    public fun run(
        initialState: S,
        onEvent: (GraphEvent) -> Unit = {},
    ): RunResult<S> {
        val endNode = requireNotNull(graph.end)
        val runtimeState =
            RuntimeState(
                workQueue = ArrayDeque(listOf(WorkItem(requireNotNull(graph.start), initialState))),
            )
        val completedStates = mutableMapOf<String, S>()

        while (true) {
            val workItem = runtimeState.dequeue()
            if (workItem != null) {
                val node = requireNotNull(graph.nodes[workItem.nodeName])
                val execution = executeNode(node, workItem.state, onEvent)
                val nextState = execution.getOrElse { cause ->
                    return fail(ExecutionError.RuntimeError(workItem.nodeName, cause), onEvent)
                }

                completedStates[workItem.nodeName] = nextState

                val childrenResult =
                    runCatching {
                        if (workItem.nodeName == endNode) {
                            emptyList()
                        } else {
                            graph.nextNodes(workItem.nodeName, nextState)
                        }
                    }
                val children = childrenResult.getOrElse { cause ->
                    return fail(ExecutionError.RuntimeError(workItem.nodeName, cause), onEvent)
                }
                val (joinNodes, readyNodes) = children.partition(graph::isMultipleParentNode)

                runtimeState.enqueue(readyNodes.map { nodeName -> WorkItem(nodeName, nextState) })
                runtimeState.addJoinArrivals(
                    joinNodes.map { joinNode ->
                        PendingJoinInput(joinNode, workItem.nodeName, nextState)
                    },
                )
                continue
            }

            if (runtimeState.hasPendingJoins) {
                val finishedJoins =
                    runtimeState.finishedJoins { nodeName -> graph.parentNodes(nodeName).toSet() }
                if (finishedJoins.isEmpty()) {
                    return fail(ExecutionError.JoinDeadlock(runtimeState.pendingJoins.keys.toSet()), onEvent)
                }

                for (nodeName in finishedJoins) {
                    val arrivals = requireNotNull(runtimeState.pendingJoins.remove(nodeName))
                    val combinedStateResult =
                        runCatching {
                            val orderedStates = graph.parentNodes(nodeName).map(arrivals::getValue)
                            orderedStates.drop(1).fold(orderedStates.first(), graph.reducer::merge)
                        }
                    val combinedState = combinedStateResult.getOrElse { cause ->
                        return fail(ExecutionError.RuntimeError(nodeName, cause), onEvent)
                    }
                    runtimeState.enqueue(listOf(WorkItem(nodeName, combinedState)))
                }
                continue
            }

            if (endNode !in completedStates) {
                return fail(ExecutionError.EndNodeNotReached(endNode), onEvent)
            }

            onEvent(GraphEvent.WorkflowCompleted(WORKFLOW_NAME))
            return RunResult.Success(completedStates.getValue(endNode))
        }
    }

    private fun executeNode(
        node: Node<S, U>,
        state: S,
        onEvent: (GraphEvent) -> Unit,
    ): Result<S> {
        val maxAttempts =
            when (val policy = node.retryPolicy) {
                RetryPolicy.Never -> 1
                is RetryPolicy.FixedAttempts -> policy.maxAttempts
            }

        repeat(maxAttempts) {
            onEvent(GraphEvent.NodeStarted(node.name))
            val attempt = runCatching { graph.reducer.reduce(state, node.action(state)) }

            if (attempt.isSuccess) {
                val nextState = attempt.getOrThrow()
                onEvent(GraphEvent.NodeCompleted(node.name))
                onEvent(GraphEvent.CheckpointCreated(node.name, nextState))
                return Result.success(nextState)
            }

            val cause = attempt.exceptionOrNull() ?: error("Failed attempt has no cause")
            onEvent(GraphEvent.NodeFailed(node.name, cause))
            if (it == maxAttempts - 1) return Result.failure(cause)
        }

        error("Retry policy produced no attempts")
    }

    private fun fail(
        error: ExecutionError,
        onEvent: (GraphEvent) -> Unit,
    ): RunResult.Failure {
        onEvent(GraphEvent.WorkflowFailed(WORKFLOW_NAME, error.cause))
        return RunResult.Failure(error)
    }

    private companion object {
        private const val WORKFLOW_NAME: String = "workflow"
    }
}
